rfd/0193-access-control-decision-api.md

```diff
@@ -0,0 +1,263 @@
+the agent-side teleport logic that previously would have worked with some combination of certificate identity and role set will instead
+take the appropriate permit structure as input. Permits will *always* be passed by pointer, not value, as a precaution against zero
+value bugs resulting in unintended access.
+
+- Methods will conventionally take the form `rpc Evaluate<Action>(<Action>Request) returns (<Action>Permit)`. For example,
+the method for evaluating server access would read `rpc EvaluateSSHAccess(SSHAccessRequest) returns (SSHAccessPermit)`.
+Note that this deviates somewhat from common PDP conventions in that we don't return a decision object wrapping the permit. Any result
+other than allow must be communicated as an error.
+
+- Common `RequestMetadata` and `PermitMetadata` types will be provided and should be included in the request and permit
+respectively. These types will contain common fields like teleport version of PDP/PEP, dry run flag, etc.
+
+- A standard `Resource` reference types with fields like  `kind` and `name` will be provided and should be prefered in
+those APIs where specifying individual resources or sets of resources is necessary. PDPs will use this reference to load
+the appropriate resource from local caches.
+
+- Two standard `Identity` types will be provided for use in communicating caller identity within decision requests. One
+corresponding to standard teleport TLS cert format, and one to the standard SSH cert format. Eventually, we would like
+the PDP to be able to make decisions based on a standardized user/subject reference similar to the `Resource`
+reference, with most fields currently stored in the certificate instead determined at runtime. In practice, moving away from
+teleport's dependence on complex certificate-based identities is non-trivial. In order to support an iterative transition, we will
+reimplement our existing identity types as a protobuf type and use that as the standard subject identifier. Individual fields will be
+transitioned from being cert-bound to being determined PDP-side in an iterative manner.
+
+Here is a truncated example of the PDP API for server access:
+
+```protobuf
+service DecisionService {
+
+  rpc EvaluateSSHAccess(SSHAccessRequest) returns (SSHAccessPermit);
+
+  // ...
+}
+
+message SSHAccessRequest {
+  RequestMetadata metadata = 1;
+
+  SSHIdentity user = 2;
+
+  Resource server = 3;
+
+  string login = 4;
+}
+
+message SSHAccessPermit {
+  PermitMetadata metadata = 1;
+
+  repeated string logins = 2;
+
+  bool forward_agent = 3;
+
+  google.protobuf.Duration max_session_ttl = 4;
+
+  bool port_forwarding = 5;
+
+  int64 client_idle_timeout = 6;
+
+  bool disconnect_expired_cert = 7;
+
+  repeated string bpf = 8;
+
+  bool x11_forwarding = 9;
+
+  int64 max_connections = 10;
+
+  // ... (there's a lot more that needs to go here)
+}
+
+message SSHIdentity {
+  string teleport_user = 1;
+  
+  string impersonator = 2;
+  
+  string login = 3;
```

Comment:
does this come from the principals in the SSH cert and if so should it be `repeated`?

Reply:
You're right that ssh certs represent logins as a repeated value.  The values here were pulled from `srv.IdentityContext` for example purposes.  Turns out that we don't really have a good single source of truth for what an ssh identity is.  I've removed all fields from the example `SSHIdentity` for now.  We're going to do some refactoring and establish a proper `tlsca.Identity` type for ssh identities, then base the `SSHIdentity` proto message off of that.